src/main/java/bio/terra/workspace/service/datareference/model/ReferenceObject.java

```diff
@@ -11,19 +10,28 @@
  * rules. ReferenceObject implementations must be fully serializable by the output of getProperties.
  */
 public interface ReferenceObject {
-  Map<String, String> getProperties();
 
   /**
-   * Convenience method for building a concrete referenceObject from a map and type.
+   * Method for serializing a ReferenceObject to a json string.
    *
-   * <p>This is used for deserializing ReferenceObjects from their properties and type.
+   * <p>This method is used for Stairway serialization. Database serialization should use the static
+   * {@Code DataReferenceDao.serializeReferenceObject} method instead.
    */
-  static ReferenceObject toReferenceObject(DataReferenceType type, Map<String, String> properties) {
+  String toJson();
+
+  /**
+   * Method for deserializing a ReferenceObject from a json string and its type.
+   *
+   * <p>This method is used for Stairway deserialization. Database deserialization should use the
+   * static {@Code DataReferenceDao.deserializeReferenceObject} method instead.
+   *
+   * <p>Implementations of this interface are expected to provide a static method for deserializing
+   * from json which should be referenced here. This should
+   */
+  static ReferenceObject fromJson(String jsonString, DataReferenceType type) {
```

Comment:
Can we remove the DataReferenceType argument and have symmetry with toJson? It'd be nice if we were able to tell the reference type from the json alone and did not need to store another field next to the json. As-is, it's funny that the serialization relies on some side piece of serialization to be deserialized later. This could take the form of a field in the json reserved for the DataReferenceType, or something else.

Reply:
It's a little weird this way, but my issue with storing the type in a json field is that every ReferenceObject implementation needs to serialize type to the same magic key inside `toJson`, and I don't know how to enforce that at the interface level. Reading that type and then appropriately deserializing in Jackson also seems really messy, and IMO that feels worse than explicitly requiring this extra information. Thoughts?

Reply:
I'm not super concerned about enforcing it at the interface level. This interface and all its impls are tiedly coupled, so we're not really likely to change them independently. These are also pretty small classes. We can & should, however, write unit tests for each impl that `assertEquals(x, fromJson(x.toJson())` which will verify we're doing whatever magic key stuff correctly.

One option for enforcing this at the interface level is to make this an abstract class, have the `public final String toJson()` on it, and a `protected abstract JsonObjectType innerToJson` that sub classes override to fill out their fields. Also have a `protected abstract DataReferenceType dataReferenceType()`. Then the base class `toJson` calls both of those protected methods to construct the right json with the magic field key. We can runtime verify that the subclasses are setting the value of the magic field key in the value returned by innerToJson.
I think it's fine to do that with more protected methods, or to just comment about magic key values needing to be filled in. Both make a cleaner public interface, which is most important. Either way, we want tests.

Reply:
Got it. For some reason I thought AutoValue interacted weirdly if you extended an abstract class, but maybe not. 

I also realized interfaces/implementations can be handled natively in Jackson though, so I think adding the field via annotation is probably the right fix here. Future implementation classes will need to update the `JsonSubTypes` annotation in `ReferenceObject`, but I think the comments on the class and `fromJson` should make that clear.

I'm leaning towards keeping the type enum as a column in the database so that we can query it in the future, but it also might be possible to do that entirely from the json object. I'm not familiar enough with json in Postgres to know what we can or can't do, though